packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -156,16 +178,28 @@ contract SignalService is AuthorizableContract, ISignalService {
         uint64 srcChainId,
         address srcApp,
         bytes32 srcSignal,
-        bytes memory merkleProof
+        bytes memory merkleProof,
+        bytes memory cachedRoot
     )
         public
         view
         virtual
+        returns (bytes memory toBeCachedStorageRoot)
     {
         address signalService = resolve(srcChainId, "signal_service", false);
 
         bytes32 slot = getSignalSlot(srcChainId, srcApp, srcSignal);
-        bool verified = LibTrieProof.verifyWithAccountProof(stateRoot, signalService, slot, hex"01", merkleProof);
+        bool verified;
+
+        if (cachedRoot.length == 0) {
+            (verified, toBeCachedStorageRoot) = LibTrieProof.verifyWithAccountProof(
+                stateRoot, signalService, slot, hex"01", merkleProof, hex""
+            );
+        } else {
+            (verified, toBeCachedStorageRoot) = LibTrieProof.verifyWithAccountProof(
+                stateRoot, signalService, slot, hex"01", merkleProof, cachedRoot
+            );
+        }
```

Comment:
Just pass in the cached root as is, let the function do what is necessary based on its value, you already handle it anyway in `LibTrieProof`.